src/test/java/hudson/plugins/throttleconcurrents/ThrottleStepTest.java

```diff
@@ -0,0 +1,122 @@
+
+import java.util.Arrays;
+import java.util.Collections;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class ThrottleStepTest {
+    private static final String ONE_PER_NODE = "one_per_node";
+    private static final String TWO_TOTAL = "two_total";
+
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public TemporaryFolder firstAgentTmp = new TemporaryFolder();
+    @Rule
+    public TemporaryFolder secondAgentTmp = new TemporaryFolder();
+
+    public void setupAgentsAndCategories() throws Exception {
+        DumbSlave firstAgent = new DumbSlave("first-agent", "dummy agent", firstAgentTmp.getRoot().getAbsolutePath(),
+                "4", Node.Mode.NORMAL, "on-agent", story.j.createComputerLauncher(null),
+                RetentionStrategy.NOOP, Collections.<NodeProperty<?>>emptyList());
+
+        DumbSlave secondAgent = new DumbSlave("second-agent", "dummy agent", secondAgentTmp.getRoot().getAbsolutePath(),
+                "4", Node.Mode.NORMAL, "on-agent", story.j.createComputerLauncher(null),
+                RetentionStrategy.NOOP, Collections.<NodeProperty<?>>emptyList());
+
+        story.j.jenkins.addNode(firstAgent);
+        story.j.jenkins.addNode(secondAgent);
+
+        ThrottleJobProperty.ThrottleCategory firstCat = new ThrottleJobProperty.ThrottleCategory(ONE_PER_NODE, 1, 0, null);
+        ThrottleJobProperty.ThrottleCategory secondCat = new ThrottleJobProperty.ThrottleCategory(TWO_TOTAL, 0, 2, null);
+
+        ThrottleJobProperty.DescriptorImpl descriptor = story.j.jenkins.getDescriptorByType(ThrottleJobProperty.DescriptorImpl.class);
+        assertNotNull(descriptor);
+        descriptor.setCategories(Arrays.asList(firstCat, secondCat));
+    }
+
+    @Test
+    public void onePerNode() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                setupAgentsAndCategories();
+                WorkflowJob firstJob = story.j.jenkins.createProject(WorkflowJob.class, "first-job");
+                // This should be sandbox:true, but when I do that, I get org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use method groovy.lang.GroovyObject invokeMethod java.lang.String java.lang.Object
+                // And I cannot figure out why. So for now...
+                firstJob.setDefinition(new CpsFlowDefinition("throttle('" + ONE_PER_NODE + "') {\n" +
+                        "  echo 'hi there'\n" +
+                        "  node('first-agent') {\n" +
+                        "    semaphore 'wait-first-job'\n" +
+                        "  }\n" +
+                        "}\n", false));
```

Comment:
Soooo - as I mentioned a few lines up, I'm getting script security exceptions on this when the sandbox is enabled and I can't figure out why, so for the moment, disabling script security until someone (probably @jglick!) tells me what I'm doing wrong to cause those errors to show up. =)